src/actuator/ActuatorFunctorsFAST.C

```diff
@@ -0,0 +1,178 @@
+}
+
+ActFastComputeForce::ActFastComputeForce(ActuatorBulkFAST& actBulk)
+  : force_(helper_.get_local_view(actBulk.actuatorForce_)),
+    offset_(helper_.get_local_view(actBulk.turbIdOffset_)),
+    turbId_(actBulk.localTurbineId_),
+    fast_(actBulk.openFast_)
+{
+  helper_.touch_dual_view(actBulk.actuatorForce_);
+}
+
+void
+ActFastComputeForce::operator()(int index) const
+{
+
+  auto pointForce = Kokkos::subview(force_, index, Kokkos::ALL);
+  const int localId = index - offset_(turbId_);
+
+  fast_.getForce(pointForce.data(), localId, turbId_);
+}
+
+ActFastSetUpThrustCalc::ActFastSetUpThrustCalc(ActuatorBulkFAST& actBulk)
+  : actBulk_(actBulk)
+{
+}
+
+void
+ActFastSetUpThrustCalc::operator()(int index) const
+{
+  auto hubLoc = Kokkos::subview(actBulk_.hubLocations_, index, Kokkos::ALL);
+  auto hubOri = Kokkos::subview(actBulk_.hubOrientation_, index, Kokkos::ALL);
+  auto thrust = Kokkos::subview(actBulk_.turbineThrust_, index, Kokkos::ALL);
+  auto torque = Kokkos::subview(actBulk_.turbineTorque_, index, Kokkos::ALL);
+
+  for (int i = 0; i < 3; i++) {
+    thrust(i) = 0.0;
+    torque(i) = 0.0;
+  }
+
+  if (actBulk_.localTurbineId_ == index) {
+    actBulk_.openFast_.getHubPos(hubLoc.data(), index);
+    actBulk_.openFast_.getHubShftDir(hubOri.data(), index);
+  } else {
+    for (int j = 0; j < 3; j++) {
+      hubLoc(j) = 0.0;
+      hubOri(j) = 0.0;
+    }
+  }
+}
+
+void
+ActFastComputeThrustInnerLoop::operator()(
+  const uint64_t pointId,
+  const double* nodeCoords,
+  double* sourceTerm,
+  const double,
+  const double scvIp) const
+{
+
+  auto offsets = actBulk_.turbIdOffset_.view_host();
+
+  // shouldn't thrust and torque contribs only come from blades?
+  // probably not worth worrying about since this is just a debug calculation
+
+  // determine turbine
+  int turbId = 0;
+  const int nPointId = static_cast<int>(pointId);
+  for (; turbId < offsets.extent_int(0); turbId++) {
+    if (nPointId >= offsets(turbId)) {
+      break;
+    }
+  }
+
+  auto hubLoc = Kokkos::subview(actBulk_.hubLocations_, turbId, Kokkos::ALL);
+  auto hubOri = Kokkos::subview(actBulk_.hubOrientation_, turbId, Kokkos::ALL);
+  auto thrust = Kokkos::subview(actBulk_.turbineThrust_, turbId, Kokkos::ALL);
+  auto torque = Kokkos::subview(actBulk_.turbineTorque_, turbId, Kokkos::ALL);
+
+  double r[3], rPerpShaft[3], forceTerm[3];
+
+  for (int i = 0; i < 3; i++) {
+    forceTerm[i] = sourceTerm[i] * scvIp;
+    r[i] = nodeCoords[i] - hubLoc(i);
+    thrust(i) += forceTerm[i];
+  }
+
+  double rDotHubOri = 0;
+  for (int i = 0; i < 3; i++) {
+    rDotHubOri += r[i] * hubOri(i);
+  }
+
+  for (int i = 0; i < 3; i++) {
+    rPerpShaft[i] = r[i] - rDotHubOri * hubOri(i);
+  }
+
+  torque(0) += (rPerpShaft[1] * forceTerm[2] - rPerpShaft[2] * forceTerm[1]);
+  torque(1) += (rPerpShaft[2] * forceTerm[0] - rPerpShaft[0] * forceTerm[2]);
+  torque(2) += (rPerpShaft[0] * forceTerm[1] - rPerpShaft[1] * forceTerm[0]);
```

Comment:
I think scvIp needs to multiply the whole term here.

Reply:
It is factored in since it is in all the components of the force term.